Non parsing errors silently fails and causes timeout. No error displayed.

**I'm submitting a ...**  (check one with "x")
```
[x ] bug report
[x ] feature request
[ ] support request => Please do not submit support request here, instead use https://groups.google.com/d/forum/nightwatch-cucumber
```

**Current behavior**
When a type error or similar occur in a custom command, step definition or assertion the test will silently fail and cause a timeout.

**Expected/desired behavior**
I should get an error message and stack trace to be able to find the issue.

**Reproduction of the problem**
If the current behavior is a bug or you can illustrate your feature request better with an example, please provide the steps to reproduce and if possible a minimal demo of the problem via https://c9.io, http://www.koding.com, https://www.nitrous.io or similar.

```
#testCommand.js
const testCommand = function() {
    var test = undefinedVar;
    return this;
};
exports.command = testCommand;

#stepDef.js
this.Given(/^Test command$/, function(client) {
    client.testCommand();
});

#test.feature
Scenario: Test command
    Given Test command
```

**What is the expected behavior?**
It should fail and output an error


**What is the motivation / use case for changing the behavior?**
It's right now really hard to find the cause of the issue. Only way is to console log you way through the file and see how far you get.


**Please tell us about your environment:**

* **nightwatch-cucumber version:**
5.6.3
* **nightwatch version:**
0.9.9
* **cucumber version:**
1.3.0
* **Node.js version:**
7.2.0

This is what is looks like when I disconnect Nightwatch-cucumber and try to run the following test

```
module.exports = {
    "Demo test Google": function(browser) {
        browser
            .openPage()
            .end();
    }
};
```

Inside I have deliberately added this: 
```
var foo = test; // Test is undefined
```

And the result in Nightwatch
```
Starting selenium server... started - PID:  1120

[Test] Test Suite
=====================

Running:  Demo test Google
Error while running openPage command: test is not defined
    at Object.Demo test Google (C:\my-path\temp\test.js:4:14)
    at Module.call (C:\my-path\node_modules\nightwatch\lib\runner\module.js:62:34)
    at C:\my-path\node_modules\nightwatch\lib\runner\testcase.js:70:29
    at _fulfilled (C:\my-path\node_modules\q\q.js:834:54)
    at self.promiseDispatch.done (C:\my-path\node_modules\q\q.js:863:30)
    at Promise.promise.promiseDispatch (C:\my-path\node_modules\q\q.js:796:13)
    at C:\my-path\node_modules\q\q.js:556:49
    at runSingle (C:\my-path\node_modules\q\q.js:137:13)
    at flush (C:\my-path\node_modules\q\q.js:125:13)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)

FAILED:  1 errors (6.147s)

 _________________________________________________

 TEST FAILURE: 1 error during execution,  0 assertions failed, 0 passed. (6.474s)

 × test

   - Demo test Google (6.147s)

   Error while running openPage command: test is not defined
       at Object.Demo test Google (C:\my-path\temp\test.js:4:14)
       at _combinedTickCallback (internal/process/next_tick.js:67:7)


Error while running openPage command: test is not defined
    at Object.Demo test Google (C:\my-path\temp\test.js:4:14)
    at Module.call (C:\my-path\node_modules\nightwatch\lib\runner\module.js:62:34)
    at C:\my-path\node_modules\nightwatch\lib\runner\testcase.js:70:29
    at _fulfilled (C:\my-path\node_modules\q\q.js:834:54)
    at self.promiseDispatch.done (C:\my-path\node_modules\q\q.js:863:30)
    at Promise.promise.promiseDispatch (C:\my-path\node_modules\q\q.js:796:13)
    at C:\my-path\node_modules\q\q.js:556:49
    at runSingle (C:\my-path\node_modules\q\q.js:137:13)
    at flush (C:\my-path\node_modules\q\q.js:125:13)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)

npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "start"
npm ERR! node v7.2.0
npm ERR! npm  v3.10.9
npm ERR! code ELIFECYCLE
npm ERR! @company/ui-test@0.2.0 start: `node src/index.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @company/ui-test@0.2.0 start script 'node src/index.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the @company/ui-test package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node src/index.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs @company/ui-test
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls @company/ui-test
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\my-path\npm-debug.log
```

--- a/lib/nightwatch-api.js
+++ b/lib/nightwatch-api.js
@@ -140,18 +140,32 @@ module.exports = class NightwatchApi {
     if (promiseInterface) return result
 
     yield new Promise((resolve, reject) => {
-      const errorHandler = (err) => {
+      let originalTerminate
+      const restore = () => {
+        this.client.get().terminate = originalTerminate
         this.client.removeListener('error', errorHandler)
+        this.client.removeListener('complete', completeHandler)
+      }
+      const errorHandler = (err) => {
+        restore()
         reject(err)
       }
       const completeHandler = (results, errors) => {
-        this.client.removeListener('complete', completeHandler)
+        restore()
         const lastError = this.client.get().results.lastError
         if (lastError || errors && errors.length) {
           lastError.stack = [lastError.message, lastError.stack].join('\n')
           return reject(lastError || new Error(errors[0]))
         }
         resolve()
+      }
+      originalTerminate = this.client.get().terminate
+      this.client.get().terminate = () => {
+        const clientError = this.client.get().errors[0]
+        const cutPos = clientError.indexOf(':')
+        const err = new Error(clientError.substring(0, cutPos))
+        err.stack = clientError
+        errorHandler(err)
       }
       this.client.on('error', errorHandler)
       this.client.on('complete', completeHandler)

--- a/test/error.test.js
+++ b/test/error.test.js
@@ -169,4 +169,28 @@ describe('Error handling', () => {
         result.exitCode.should.equal(1)
       })
   })
+
+  it('should handle errors in custom commands', () => {
+    return testCaseFactory
+      .create('customCommandsTest')
+      .customCommand('testCommand', `module.exports.command = function () {
+    var test = undefinedVar;
+    return this;
+}`)
+      .feature('addition')
+      .scenario('small numbers')
+      .given('User is on the simple calculator page', function () { this.init() })
+      .and('User enter 4 in A field', function () { this.testCommand() })
+      .and('User enter 5 in B field', function () { this.page.calculator().setValue('@numberB', 5) })
+      .when('User press Add button', function () { this.page.calculator().click('@addButton') })
+      .then('The result should contain 9', function () { this.page.calculator().assert.containsText('@result', 9) })
+      .run()
+      .then((result) => {
+        result.features[0].result.status.should.be.failed
+        result.features[0].result.scenarioCounts.should.deep.equal({failed: 1})
+        result.features[0].scenarios[0].result.status.should.be.failed
+        result.features[0].scenarios[0].result.stepCounts.should.deep.equal({skipped: 3, failed: 1, passed: 1})
+        result.output.should.contain('Error while running testCommand command: undefinedVar is not defined')
+      })
+  })
 })

--- a/test/fixture/nightwatch.conf.js.tmpl
+++ b/test/fixture/nightwatch.conf.js.tmpl
@@ -30,7 +30,9 @@ module.exports = {
   src_folders: ['../../test/fixture/plain-nightwatch-test'],
   <% } %>
   output_folder: 'reports',
-  custom_commands_path: '',
+  <% if (customCommands) { %>
+  custom_commands_path: 'custom_commands',
+  <% } %>
   custom_assertions_path: '',
   <% if (pageObjects) { %>
   page_objects_path: 'page_objects',

--- a/test/test-case-factory.js
+++ b/test/test-case-factory.js
@@ -30,6 +30,7 @@ class TestCaseFactory {
     this.groups = []
     this.stepDefinitions = []
     this.pageObjects = []
+    this.customCommands = []
     this.currentGroup = {
       features: {}
     }
@@ -117,6 +118,11 @@ class TestCaseFactory {
     return this
   }
 
+  customCommand (name, source) {
+    this.customCommands.push({name, source})
+    return this
+  }
+
   _buildExamples (featureFile, scenario) {
     const maxColWidths = []
     fs.writeFileSync(featureFile, `\n  Examples:\n`, { flag: 'a' })
@@ -192,8 +198,18 @@ class TestCaseFactory {
     })
   }
 
+  _buildCustomCommands () {
+    if (!this.customCommands.length) return
+
+    mkdirp.sync(path.join(this.testCasePath, 'custom_commands'))
+    this.customCommands.forEach((customCommand) => {
+      fs.writeFileSync(path.join(this.testCasePath, 'custom_commands', `${customCommand.name}.js`), customCommand.source)
+    })
+  }
+
   _build () {
     this.options.pageObjects = !!this.pageObjects.length
+    this.options.customCommands = !!this.customCommands.length
     this.testCasePath = path.join(process.cwd(), 'tmp', this.name)
     mkdirp.sync(this.testCasePath)
     fs.writeFileSync(path.join(this.testCasePath, 'nightwatch.conf.js'), nightwatchConfTemplate(this.options))
@@ -208,6 +224,7 @@ class TestCaseFactory {
 
     this._buildStepDefinitions()
     this._buildPageObjects()
+    this._buildCustomCommands()
 
     let groupPath
     this.groups.forEach((group) => {

